python/pyarrow/tests/test_extension_type.py

```diff
@@ -517,10 +517,31 @@ def test_cast_kernel_on_extension_arrays():
     assert isinstance(casted, pa.ChunkedArray)
 
 
-def test_casting_to_extension_type_raises():
-    arr = pa.array([1, 2, 3, 4], pa.int64())
-    with pytest.raises(pa.ArrowNotImplementedError):
-        arr.cast(IntegerType())
+@pytest.mark.parametrize("data,ty", (
+    ([1, 2], pa.int32),
+    ([1, 2], pa.int64),
+    (["1", "2"], pa.string),
+    ([b"1", b"2"], pa.binary),
+    ([1.0, 2.0], pa.float32),
+    ([1.0, 2.0], pa.float64)
+))
+def test_casting_to_extension_type(data, ty):
+    arr = pa.array(data, ty())
+    out = arr.cast(IntegerType())
+    assert isinstance(out, pa.ExtensionArray)
+    assert out.type == IntegerType()
+    assert out.to_pylist() == [1, 2]
```

Comment:
I didn't check if this is covered by the C++ tests, but we should also add some failures cases (where casting to the storage type fails, either always or depending on an option, eg disallowing vs allowing float truncation when casting a float array to this IntegerType)

Reply:
There are indeed some C++ testing, such as `IntToExtensionTypeDowncast` (works only when setting options to allow truncation) and `PrimitiveToExtension` (type which is impossible to cast)

Happy to add them to Python as well.